src/gas_metering/mod.rs

```diff
@@ -484,23 +515,58 @@ pub fn inject<R: Rules>(raw_wasm: &[u8], rules: &R, gas_module_name: &str) -> Re
     // We'll push the gas counter fuction after all other functions
     let gas_func = total_func;
 
-    // grow_cnt_func is optional, so we put it after the gas function which always exists
-    let grow_cnt_func = total_func + 1;
-
-    let mut need_grow_counter = false;
     let mut error = false;
 
+    // Read types which are needed in later steps
+    let mut functype_param_counts = Vec::new();
+    if let Some(type_section) = module_info.raw_sections.get_mut(&SectionId::Type.into()) {
+        let type_sec_reader = TypeSectionReader::new(&type_section.data, 0)?;
+
+        for t in type_sec_reader {
+            let Type::Func(ft) = t?;
+            let count = ft.params().len() as u32;
+
+            functype_param_counts.push(count);
+        }
+    }
```

Comment:
FYI, you could have written:

```rust
    // Read types which are needed in later steps
    let functype_param_counts = module_info
        .raw_sections
        .get_mut(&SectionId::Type.into())
        .map(|type_section| TypeSectionReader::new(&type_section.data, 0))
        .transpose()?
        .map(|reader| reader.into_iter())
        .into_iter()
        .flatten()
        .map(|t| t.map(|Type::Func(ft)| ft.params().len() as u32))
        .collect::<Result<Vec<_>, _>>()?;
```

I'm glad you didn't. But the rust police made me say this.

Reply:
I did come up with something resembling this, but uhh.. I like readable code